KAFKA-5894: add the notion of max inflight requests to async ZooKeeperClient

onurkaraman opened the pull request:
ZooKeeperClient is a zookeeper client that encourages pipelined requests to zookeeper. We want to add the notion of max inflight requests to the client for several reasons:
1. to bound memory overhead associated with async requests on the client.
2. to not overwhelm the zookeeper ensemble with a burst of requests.

tedyu commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -36,13 +36,18 @@ import org.apache.zookeeper.{CreateMode, WatchedEvent, Watcher, ZooKeeper}
+                      sessionTimeoutMs: Int,
+                      connectionTimeoutMs: Int,
+                      stateChangeHandler: StateChangeHandler,
+                      maxInFlightRequests: Int = Int.MaxValue) extends Logging {
```
Where is the lower limit specified ?

onurkaraman replied:
Hi @tedyu. What do you mean by lower limit?

tedyu replied:
By default, there is no limit.
Who would be enforcing the limit ?

onurkaraman replied:
I'm still not sure I understand your question.

The client's maxInFlightRequests is configurable.

If you want the feature, you'd set the limit when instantiating the client.
If you don't want the feature, it's set to `Int.MaxValue` by default which effectively means there is no limit.

The actual limit is enforced through the `inFlightRequests` semaphore which increments when a request goes over the wire and decrements when we get a response back.

tedyu replied:
Is it possible to add unit test which imposes lower limit and verifies that this works ?

Thanks for your patience.

tedyu replied:
The release of semaphore and decrement of countDownLatch are paired.
Consider creating a method which does these two operations.

junrao replied:
Could we get rid of the default?

omkreddy commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -36,13 +36,18 @@ import org.apache.zookeeper.{CreateMode, WatchedEvent, Watcher, ZooKeeper}
   * @param connectionTimeoutMs connection timeout in milliseconds
   * @param stateChangeHandler state change handler callbacks called by the underlying zookeeper client's EventThread.
   */
```
need to add maxInFlightRequests param to javadoc

junrao reviewed (commented):
@onurkaraman : Thanks for the patch. Looks good overall. If we want to make maxInFlightRequests configurable, we will need a KIP.

onurkaraman commented:
Hey @junrao. I just rebased this PR.

junrao commented:
@onurkaraman : Thanks for the rebasing. Looks good overall. Do you want to add a new config for the max inflight request in KafkaConfg and hook it up to ZookeeperClient?

onurkaraman commented:
I rebased again, added the `zookeeper.max.in.flight.requests` config, and wired it up.

I'm running the experiments from https://github.com/apache/kafka/pull/3765 but varying this config value to measure the impact and find a reasonable default value.

onurkaraman commented:
I reran the controlled shutdown test from https://github.com/apache/kafka/pull/3765 but varied the `zookeeper.max.in.flight.requests` config. Here are the results:
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=1`:
```
^C[2017-10-20 06:34:32,272] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 06:34:32,277] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 06:34:32,278] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 06:34:42,011] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 06:34:47,013] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=2`:
```
^C[2017-10-20 07:21:10,951] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:21:10,955] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:21:10,956] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:21:17,137] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:21:21,267] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=5`:
```
^C[2017-10-20 07:27:59,684] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:27:59,688] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:27:59,689] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:28:04,268] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:28:08,908] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=10`:
```
^C[2017-10-20 07:12:10,020] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:12:10,027] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:12:10,028] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:12:13,700] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:12:18,017] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=100`:
```
^C[2017-10-20 07:42:15,281] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:42:15,287] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:42:15,288] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:42:18,515] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:42:22,442] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=1000`:
```
^C[2017-10-20 07:01:01,979] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:01:01,985] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:01:01,986] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:01:05,121] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:01:09,870] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=2147483647`:
```
^C[2017-10-20 07:35:54,274] INFO Terminating process due to signal SIGINT (kafka.Kafka$)
[2017-10-20 07:35:54,277] INFO [KafkaServer id=1] shutting down (kafka.server.KafkaServer)
[2017-10-20 07:35:54,278] INFO [KafkaServer id=1] Starting controlled shutdown (kafka.server.KafkaServer)
[2017-10-20 07:35:57,372] INFO [KafkaServer id=1] Controlled shutdown succeeded (kafka.server.KafkaServer)
[2017-10-20 07:36:01,851] INFO [KafkaServer id=1] shut down completed (kafka.server.KafkaServer)
```
So it seems we can actually set the config pretty low (maybe 10) and get most of the benefits.

Side note: One interesting note is that KAFKA-5894 controlled shutdown with `zookeeper.max.in.flight.requests=1` performed better than the trunk runs from https://github.com/apache/kafka/pull/3765 prior to KAFKA-5642. I dug a bit into this and found two causes:
1. controlled shutdown times prior to KAFKA-5642 appeared to have much more variance.
2. tweaking KAFKA-5894 to perform controlled shutdown with zero batching (both in the state machine calls and in the calls to `ControllerBrokerRequestBatch.sendRequestsToBrokers`) with `zookeeper.max.in.flight.requests=1` increases the controlled shutdown time, so some improvements came from more effectively using ControllerBrokerRequestBatch.

onurkaraman commented:
Updated the PR to set the default config value to 10.

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -418,6 +419,7 @@ object KafkaConfig {
   val ZkConnectionTimeoutMsDoc = "The max time that the client waits to establish a connection to zookeeper. If not set, the value in " + ZkSessionTimeoutMsProp + " is used"
   val ZkSyncTimeMsDoc = "How far a ZK follower can be behind a ZK leader"
   val ZkEnableSecureAclsDoc = "Set client to use secure ACLs"
+  val ZkMaxInFlightRequestsDoc = "The maximum number of unacknowledged requests the client will send before blocking."
```
the client will send => the client will send to Zookeeper?

onurkaraman replied:
Done.

junrao reviewed (commented):
@onurkaraman : Thanks for the patch and the experimental results. Just a couple of minor comments. Also, waiting for the KIP to go through.

omkreddy commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -691,6 +693,7 @@ object KafkaConfig {
       .define(ZkSyncTimeMsProp, INT, Defaults.ZkSyncTimeMs, LOW, ZkSyncTimeMsDoc)
       .define(ZkEnableSecureAclsProp, BOOLEAN, Defaults.ZkEnableSecureAcls, HIGH, ZkEnableSecureAclsDoc)
+      .define(ZkMaxInFlightRequestsProp, INT, Defaults.ZkMaxInFlightRequests, HIGH, ZkMaxInFlightRequestsDoc)
 
```
can we add a atLeast(1) validator here?

onurkaraman replied:
Makes sense.

omkreddy commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -74,50 +80,64 @@ class ZookeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
+
+  private def send(request: AsyncRequest, countDownLatch: CountDownLatch, responseQueue: ArrayBlockingQueue[AsyncResponse]): Unit = {
+    import scala.collection.JavaConverters._
+    inFlightRequests.acquire()
```
Normally we use try..finally.. block to release the semaphore permit.  otherwise, we may not be releasing the semaphore permits on unexpected errors.

onurkaraman committed:
KAFKA-5894: add the notion of max inflight requests to async ZooKeeperClient

ZooKeeperClient is a zookeeper client that encourages pipelined requests to zookeeper. We want to add the notion of max inflight requests to the client for several reasons:
1. to bound memory overhead associated with async requests on the client.
2. to not overwhelm the zookeeper ensemble with a burst of requests.

onurkaraman commented:
Updated the PR after rebasing against trunk.

junrao commented:
@onurkaraman : Thanks for the patch. LGTM. Waiting for the KIP vote to pass before merging.

omkreddy commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -81,8 +86,10 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
         send(request) { response =>
           responseQueue.add(response)
+          inFlightRequests.release()
           countDownLatch.countDown()
```
incase you missed my previous comment,  I just want to remind that we will miss releasing semaphore permits incase of validation errors from zookeeper client. I agree that, this may not be a problem,  if we have enough validation in KafkaZkClient. sample test case given below. 

```java
  @Test
  def testSemaphorePermitLeak(): Unit = {
    import scala.collection.JavaConverters._
    val data = bytes
    var zooKeeperClient: ZooKeeperClient = null
    var createResponse: CreateResponse = null

    try {
      zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, 1, null)
      createResponse = zooKeeperClient.handleRequest(CreateRequest("illegalpath", data, ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala,
        CreateMode.PERSISTENT))
    } catch {
      case e: Throwable => info("Exception occured: " + e.getMessage)
    }
    assertEquals("Permit should be 1 ", 1, zooKeeperClient.inFlightRequests.availablePermits())
  }
```

onurkaraman replied:
Hi @omkreddy. I did see your earlier comment.

While I knew that you should use try-finally with these concurrency primitives to make sure you don't get unexpectedly stuck, I had trouble finding where the problem would actually occur.

Your follow-up comment helped though was still not totally specific. More specifically, `org.apache.zookeeper.ZooKeeper` async apis internally call `org.apache.zookeeper.common.PathUtils.validatePath`, which throws `IllegalArgumentException` if the path is invalid based on some local validation checks.

In which case, I think the change would be the following:
```
requests.foreach { request =>
  inFlightRequests.acquire()
  try {
    send(request) { response =>
      responseQueue.add(response)
      inFlightRequests.release()
      countDownLatch.countDown()
    }
  } finally {
    inFlightRequests.release()
  }
}
```
Is this the scenario you had in mind? Are there others? I agree that this change is needed.

Things get more complicated if we don't trust `org.apache.zookeeper.ZooKeeper` to always deliver responses back, suggesting that the semaphore and countdown latch would never get updated. I'm not sure if we need to address that here.

tedyu replied:
In the above snippet, inFlightRequests.release() is called twice.

onurkaraman replied:
Good catch @tedyu.

omkreddy replied:
@onurkaraman Yes, this is the scenario, i was thinking. 
I agree that your second scenario is bit more complex to implement.  we may need to add zk response timeouts.  If needed , we can fix in the future.

onurkaraman replied:
Ah. The failing tests reminded me of why this was more complicated than what it initially seemed. We actually want to update both the `inFlightRequests.release()` and `countDownLatch.countDown()` in the callback in the normal case, not the calling thread.

We only want to `inFlightRequests.release()` in the calling thread if the calling thread experiences an exception, in which case the calling thread won't even need to `countDownLatch.countDown()` since the calling thread wouldn't have even begun to wait on the countdown latch.

onurkaraman committed:
update concurrency primitives in finally block

onurkaraman committed:
add max in-flight requests config to KafkaZkClientTest

onurkaraman committed:
to react to exceptions, only release inFlightRequests in a try-catch

ijuma commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -81,9 +86,17 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
+        try {
+          send(request) { response =>
+            responseQueue.add(response)
+            inFlightRequests.release()
```
We should probably add a comment that no method that can throw an exception can be called after `inFlightRequests.release()` or else we will release twice.

tedyu replied:
A boolean variable can be introduced which records whether inFlightRequests has been released.
This boolean is checked in the finally block.

onurkaraman replied:
@tedyu it seems like the current PR solution is simpler.

tedyu replied:
Using the boolean, even if any call, following inFlightRequests.release(), throws exception, there wouldn't be double release.

onurkaraman replied:
@tedyu the short answer is I don't believe your flag approach works. What's proposed in the PR is correct.

Something that's easy to forget is that the following block is processed asynchronously from the `org.apache.zookeeper.EventThread`:
```
{ response =>
  responseQueue.add(response)
  inFlightRequests.release()
  countDownLatch.countDown()
}
```
On the other hand, the rest of the try/catch block in `handleRequests`, or your proposed try/finally block, is processed by the calling thread.

I think what you're suggesting is:
```
flag = false
requests.foreach { request =>
  inFlightRequests.acquire()
  try {
    send(request) { response =>
      responseQueue.add(response)
      inFlightRequests.release()
      countDownLatch.countDown()
      flag = true
    }
  } finally {
    if (!flag)
      inFlightRequests.release()
  }
}
```
In this snippet, the calling thread will quickly run through the whole try block quickly since it's asynchronous, execute the finally block, and check the flag you're proposing. By this time, the flag may or may not be set depending on how quickly zookeeper processes our request and our asynchronous block gets processed.

The behavior we actually want is split into two scenarios:
1. The happy case: the underlying `org.apache.zookeeper.ZooKeeper` client is able to successfully send out a request asynchronously to ZooKeeper and the `org.apache.zookeeper.EventThread` processes the asynchronous block after retrieving a response. We should only update the concurrency primitives after retrieving the response.
2. The sad case: the underlying `org.apache.zookeeper.ZooKeeper` client immediately throws an exception from the calling thread before even sending out a request to ZooKeeper, in which case we should only call `inFlightRequests.release`.

tedyu replied:
Makes sense.

onurkaraman replied:
@ijuma regarding https://github.com/apache/kafka/pull/3860#discussion_r148135333

I don't think a double-release is possible. We either:
1. execute the asynchronous callback block from the `org.apache.zookeeper.EventThread`, potentially resulting in an exception which the `org.apache.zookeeper.EventThread` would catch.
2. execute the catch block when the calling thread attempts to `send`.

The only risk I see is if an exception somehow gets thrown from the asynchronous block before the call to `inFlightRequests.release`, not after. In this scenario, the semaphore would not get decremented for the problematic response callback.

ijuma replied:
Thanks for the explanation. So the release in the callback should be in a try/finally?

ijuma commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -81,9 +86,17 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
       val responseQueue = new ArrayBlockingQueue[Req#Response](requests.size)
```
Not related to this PR, but out of curiosity, why are we using a blocking queue here?

onurkaraman replied:
You might be right. It doesn't look like a BlockingQueue is necessary because `CountDownLatch.await` makes the `happens-before` guarantee:
> Memory consistency effects: Until the count reaches zero, actions in a thread prior to calling countDown() happen-before actions following a successful return from a corresponding await() in another thread.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html

Should this be addressed in this PR or later?

ijuma replied:
Either way works for me.

onurkaraman replied:
Cool. I'll do it in a different PR.

ijuma commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -41,31 +41,31 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
   @Test(expected = classOf[UnknownHostException])
   def testUnresolvableConnectString(): Unit = {
-    new ZooKeeperClient("some.invalid.hostname.foo.bar.local", -1, -1, null)
+    new ZooKeeperClient("some.invalid.hostname.foo.bar.local", -1, -1, Int.MaxValue, null)
```
Shouldn't we add some tests where `zookeeper.max.in.flight.requests` is a low enough number to have an impact on the number of concurrent requests?

ijuma reviewed (commented):
Thanks for the PR. Some comments/questions.

junrao commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -81,9 +86,17 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
+            countDownLatch.countDown()
+          }
+        } catch {
+          case e: Throwable =>
```
Hmm, will a ZK operation with callback ever throw an exception?

onurkaraman replied:
If you're talking about an exception being thrown not from the callback but from the user thread calling:
- `zooKeeper.exists`
- `zooKeeper.getData`
- `zooKeeper.getChildren`
- `zooKeeper.create`
- `zooKeeper.setData`
- `zooKeeper.delete`
- `zooKeeper.getACL`
- `zooKeeper.setACL`

then the answer is yes. I described how in an earlier comment here:
https://github.com/apache/kafka/pull/3860#discussion_r148075028

> More specifically, `org.apache.zookeeper.ZooKeeper` async apis internally call `org.apache.zookeeper.common.PathUtils.validatePath`, which throws `IllegalArgumentException` if the path is invalid based on some local validation checks.

junrao reviewed (commented):
@onurkaraman : Thanks for the patch. Just one more clarification question.

onurkaraman committed:
add a test to test that requests terminate after connection loss

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

junrao commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -331,9 +332,42 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
+  @Test
+  def testConnectionLossRequestTermination(): Unit = {
+    val batchSize = 10
+    val zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, Int.MaxValue, null)
```
Could we set inflight request to sth less than the batchSize to make sure the logic still works?

junrao commented on core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:
```diff
@@ -33,7 +33,7 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
   override def setUp() {
     super.setUp()
-    zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, null)
+    zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, Int.MaxValue, null)
```
Should we set the inflight request to the default config?

junrao reviewed (commented):
@onurkaraman : Thanks for the latest patch. LGTM. Just a couple of minor comments. You can merge the patch after addressing those.

onurkaraman committed:
tweak test config values

onurkaraman commented:
Thanks for the reviews. Rebased against trunk to resolve merge conflicts.

onurkaraman commented:
retest this please

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 8051 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8053 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8053 tests run, 5 skipped, 0 failed.
--none--
